### PR TITLE
amp-bind: Support special chars in strings

### LINKED ISSUE #17863: amp-bind escapes new lines in strings
Inconsistent string escaping behavior with amp-bind. This will escape the new line char:
```
<button type="button"
  on="tap:AMP.setState({formState: {message: formState.message + '\nSome another text' } })">
  Button
</button>
```
but this wont:

```
<amp-state id="formState">
  <script type="application/json">
  {
    "message": "Some text",
    "newLine": "\n"
  }
  </script>
</amp-state>

 on="tap:AMP.setState({formState: {message: formState.message + formState.newLine + 
```
see also: https://stackoverflow.com/questions/52136567/how-to-dynamically-add-a-new-line-to-the-string-stored-in-the-amp-state

//cc @choumx

### LINKED ISSUE #22265: [amp-bind] Escaped characters in amp-state strings fail to be decoded
According to [docs](https://amp.dev/documentation/components/amp-bind#state), `amp-state` is populated with JSON data.

Nevertheless, we've seen reports in the AMP WordPress plugin where valid JSON that includes Unicode-escaped characters fail (or any escaped characters) to be decoded in AMP:

* https://github.com/ampproject/amp-wp/issues/2286:

> screenshot 10

* https://github.com/ampproject/amp-wp/issues/1791:

> screenshot 2019-01-02 at 16 16 30

We've been fixing the issue thus far by encoding JSON in PHP with the `JSON_UNESCAPED_UNICODE` flag. However, the requirement of not using escaped Unicode is not specified in the `amp-bind` docs. 

Additionally, in PHP the `/` character is automatically escaped as `\/` for security, but this is failing to be decoded in AMP. We should not advise PHP to use `JSON_UNESCAPED_SLASHES` to get around this.

So, JSON-escaped characters should be properly decoded in `amp-state` and in `on` expressions.

# Example

```html
<!doctype html>
<html ⚡>
  <head>
    <meta charset="utf-8">
    <title>amp-bind with escaped unicode</title>
    <link rel="canonical" href="self.html" />
    <meta name="viewport" content="width=device-width,minimum-scale=1">
    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
    <script async src="https://cdn.ampproject.org/v0.js"></script>
    <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>    
  </head>
  <body style="padding:20px;">
      <p [text]="text">\o/ Hello World</p>
      <button on='tap:AMP.setState({ text: "\\o\/ \u041f\u0440\u0438\u0432\u0435\u0442, \u043c\u0438\u0440!" })'>Translate to Russian</button>
  </body>
</html>
```

# Expected

> image

# Actual

> image

AMP v1904301721170

### PR DESCRIPTION
**Pull request by choumx**
Fixes #17863 and fixes #22265.

### CODE REVIEW

File: extensions/amp-bind/0.1/test/test-bind-expression.js [hunk 1]
```diff
@@ -205,6 +205,17 @@ describe('BindExpression', () => {
       expect(evaluate('+"1"')).to.equal(1);
     });
 
+    it('should parse special characters', () => {
+      expect(evaluate('"\\n"')).to.equal('\n');
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/test/test-bind-expression.js]: What does this look like when written in HTML?

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 2]
```diff
@@ -279,7 +279,9 @@ literal:
 primitive:
     STRING
       %{
-        const string = yytext.substr(1, yyleng - 2);
+        // Replace leading/trailing single-quote with double-quote chars and
+        // use JSON.parse() to process special chars e.g. '\n'.
+        const string = parseJson(`"${yytext.substr(1, yyleng - 2)}"`);
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Is this safe? What if we used a single quote and the string includes double quotes inside it?

File: extensions/amp-bind/0.1/test/test-bind-expression.js (hunk 1 above)
Comment (choumx) [reply extensions/amp-bind/0.1/test/test-bind-expression.js]: Looks like `"\n"`.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 2 above)
Comment (choumx) [reply extensions/amp-bind/0.1/bind-expr-impl.jison]: Attribute values can't have double quotes.

Comment (jridgewell) [reply extensions/amp-bind/0.1/bind-expr-impl.jison]: Note, this will also transform unicode escapes (`\u0000`) into their unicode chars.

Comment (choumx) [reply extensions/amp-bind/0.1/bind-expr-impl.jison]: Yea, that's #22265.

Comment (jridgewell) [reply extensions/amp-bind/0.1/bind-expr-impl.jison]: `<div attr='foo"bar'>` is legal HTML

Comment (choumx) [reply extensions/amp-bind/0.1/bind-expr-impl.jison]: Excellent catch. 👍

**Pull request merged**